pennylane/compiler/python_compiler/inspection/pydot_dag_builder.py

```diff
@@ -0,0 +1,205 @@
+            to_node_id (str): The unique ID of the destination node.
+            **edge_attrs (Any): Any additional styling keyword arguments.
+
+        """
+        # Use ChainMap so you don't need to construct a new dictionary
+        edge_attrs = ChainMap(edge_attrs, self._default_edge_attrs)
+        edge = pydot.Edge(from_node_id, to_node_id, **edge_attrs)
+        self.graph.add_edge(edge)
+
+    def add_cluster(
+        self,
+        cluster_id: str,
+        cluster_label: str,
+        parent_graph_id: str | None = None,
+        **cluster_attrs: Any,
+    ) -> None:
+        """Add a single cluster to the graph.
+
+        A cluster is a specific type of subgraph where the nodes and edges contained
+        within it are visually and logically grouped.
+
+        Args:
+            cluster_id (str): Unique cluster ID to identify this cluster.
+            cluster_label (str): The text to display on the cluster when rendered.
+            parent_graph_id (str | None): Optional ID of the cluster this cluster belongs to.
+            **cluster_attrs (Any): Any additional styling keyword arguments.
+
+        """
+        # Use ChainMap so you don't need to construct a new dictionary
+        cluster_attrs = ChainMap(cluster_attrs, self._default_cluster_attrs)
+        cluster = pydot.Cluster(graph_name=cluster_id, **cluster_attrs)
+
+        # Puts the label in a node within the cluster.
+        # Ensures that any edges connecting nodes through the cluster
+        # boundary don't block the label.
+        # ┌───────────┐
+        # │ ┌───────┐ │
+        # │ │ label │ │
+        # │ └───────┘ │
+        # │           │
+        # └───────────┘
+        if cluster_label:
+            node_id = f"{cluster_id}_info_node"
+            rank_subgraph = pydot.Subgraph()
+            node = pydot.Node(
+                node_id,
+                label=cluster_label,
+                shape="rectangle",
+                style="dashed",
+                fontname="Helvetica",
+                penwidth=2,
+            )
+            rank_subgraph.add_node(node)
+            cluster.add_subgraph(rank_subgraph)
+            cluster.add_node(node)
+
+        self._subgraphs[cluster_id] = cluster
+
+        parent_graph_id = "__base__" if parent_graph_id is None else parent_graph_id
+        self._subgraphs[parent_graph_id].add_subgraph(cluster)
+
+    def render(self, output_filename: str) -> None:
+        """Render the graph to a file
+
+        The implementation should ideally infer the output format
+        (e.g., 'png', 'svg') from this filename's extension.
+
+        Args:
+            output_filename (str): Desired filename for the rendered graph. If no file extension is
+                provided, it will default to a `.png` file.
+
+        """
+        format = output_filename.split(".")[-1].lower()
+        if not format:
+            format = "png"
+            output_filename += ".png"
```

Comment:
This won't work if I wanted to name my file something like `prototype.trial1.png` but I only provided `prototype.trial1` for `output_filename`. I don't mind adding a precondition for this as long as it is documented.